oracle-instantclient: 12.2.0.1.0 -> 19.3.0.0.0

flokli opened the pull request:
##### Motivation for this change
It's now possible to download Oracle Instant Client without having to
click a checkbox to agree to the license, and then log in with an Oracle
account to start the download, at least on Linux (for now).

Darwin still seems to be at 18.1 (and needs manual fetching).
Version numbers haven't been that aligned recently, so I refactored the derivation to support different version numbers and fetchers per platform.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Repro:

```sh
$(nix-build -E "with import /path/to/checkout {};python3.withPackages (p: [p.cx_oracle p.ptpython])")/bin/ptpython
```
```py
import cx_Oracle
connection = cx_Oracle.connect("MAIN_USER/MAIN_PASSWORD@localhost/orclpdb1")
```

flokli committed:
odpic: 3.1.0 -> 3.2.1

flokli committed:
oracle-instantclient: 12.2.0.1.0 -> 19.3.0.0.0

It's now possible to download Oracle Instant Client without having to
click a checkbox to agree to the license, and then log in with an Oracle
account to start the download, at least on Linux (for now).

Darwin still seems to be at 18.1, and version numbers haven't been that
aligned recently, so allow different version numbers per platform.

flokli committed:
oracle-instantclient: turn into a multi-output derivation

flokli committed:
odpic: add to rpath

flokli committed:
python3.pkgs.cx_oracle: 7.1.3 -> 7.2.2

aanderse commented on pkgs/development/libraries/oracle-instantclient/default.nix:
```diff
@@ -1,71 +1,130 @@
     # PECL::oci8 will not build without this
     # this symlink only exists in dist zipfiles for some platforms
-    ln -sfn $out/lib/libclntsh${extLib}.12.1 $out/lib/libclntsh${extLib}
+    ln -sfn $lib/lib/libclntsh${extLib}.12.1 $lib/lib/libclntsh${extLib}
```
`12.1` would need to be updated here I imagine.

flokli replied:
Not sure. The package ships multiple versions:

```
/nix/store/a9bwgf3w7xq1019kacjbwkfssp1g7n7y-oracle-instantclient-19.3.0.0.0-lib/lib/libclntsh.so
/nix/store/a9bwgf3w7xq1019kacjbwkfssp1g7n7y-oracle-instantclient-19.3.0.0.0-lib/lib/libclntsh.so.10.1
/nix/store/a9bwgf3w7xq1019kacjbwkfssp1g7n7y-oracle-instantclient-19.3.0.0.0-lib/lib/libclntsh.so.11.1
/nix/store/a9bwgf3w7xq1019kacjbwkfssp1g7n7y-oracle-instantclient-19.3.0.0.0-lib/lib/libclntsh.so.12.1
/nix/store/a9bwgf3w7xq1019kacjbwkfssp1g7n7y-oracle-instantclient-19.3.0.0.0-lib/lib/libclntsh.so.18.1
/nix/store/a9bwgf3w7xq1019kacjbwkfssp1g7n7y-oracle-instantclient-19.3.0.0.0-lib/lib/libclntsh.so.19.1
/nix/store/a9bwgf3w7xq1019kacjbwkfssp1g7n7y-oracle-instantclient-19.3.0.0.0-lib/lib/libclntshcore.so.19.1
```

I think the better fix would be to get PECL::oci8 to `dlopen`/link against the version it works with.

flokli replied:
building on current master and `readelf -d` at the `oci8.so` shows it only needs the followin libraries:

```
 0x0000000000000001 (NEEDED)             Shared library: [libclntsh.so.12.1]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
```

so we can drop the `libclntsh.so` symlink.

However, there's another problem in `phpPackages.oci8` - it didn't handle oracle-instantclient being a split derivation. Will upload a patch fixing this.

aanderse reviewed (changes requested):
Will test once I get a chance.

flokli committed:
oracle-instantclient: drop PECL::oci8 hack

This wasn't about PECL::oci8 in first place, but DBD-Oracle.

In the current version of oracle-instantclient, a libclntsh.so is
shipped, so DBD-Oracle is happy.

flokli committed:
oracle-instantclient: use copy instead of install for libraries

libclntsh.so* and libocci.so* are symlinks to the latest version.

By using `install`, these are all dereferenced and duplicated.

Let's avoid that, and shrink the lib output from 621M to 226M.

flokli committed:
perlPackages.DBDOracle: 1.76 -> 1.80

flokli commented:
Was able to build `perlPackages.DBDOracle`, `phpPackages.oci8` and `python3.pkgs.cx_oracle` successfully.

Poked at the .so files on linux, and tried connecting through the python bindings. Looks good so far :-)

@aanderse, could you take a look at situation on darwin? At least the python bindings previously did work previously…

aanderse reviewed (changes requested):
After making the mentioned changes to your branch I was able to compile the following on `darwin`:
- `oracle-instantclient`
- `php72Packages.oci8`
- `php73Packages.oci8`
- `python3.pkgs.cx_oracle`

And without making any changes I was able to compile the following on `linux`:
- `oracle-instantclient`
- `php72Packages.oci8`
- `php73Packages.oci8`
- `python3.pkgs.cx_oracle`
- `perlPackages.DBDOracle`
- `perldevelPackages.DBDOracle`

`perldevelPackages.DBDOracle` had the following failure on `darwin`:
```
#   Failed test 'require DBDOracleTestLib;'
#   at t/00dbdoracletestlib.t line 11.
#     Tried to require 'DBDOracleTestLib'.
#     Error:  Can't load '/private/var/folders/3y/f246ctxd07g_ght97sqhlhs80000gp/T/nix-build-perl5.31.2-DBD-Oracle-1.80.drv-0/DBD-Oracle-1.76/blib/arch/auto/DBD/Oracle/Oracle.bundle' for module DBD::Oracle: dlopen(/private/var/folders/3y/f246ctxd07g_ght97sqhlhs80000gp/T/nix-build-perl5.31.2-DBD-Oracle-1.80.drv-0/DBD-Oracle-1.76/blib/arch/auto/DBD/Oracle/Oracle.bundle, 2): Library not loaded: @rpath/libclntsh.dylib.18.1
#   Referenced from: /private/var/folders/3y/f246ctxd07g_ght97sqhlhs80000gp/T/nix-build-perl5.31.2-DBD-Oracle-1.80.drv-0/DBD-Oracle-1.76/blib/arch/auto/DBD/Oracle/Oracle.bundle
#   Reason: image not found at /nix/store/q9dgm21xc01as964s5fj5kqyq9cs0ivs-perl-5.31.2/lib/perl5/5.31.2/darwin-2level/DynaLoader.pm line 197.
#  at t/lib/DBDOracleTestLib.pm line 29.
```

flokli committed:
perlPackages.DBDOracle: fix darwin build

flokli commented on pkgs/development/libraries/oracle-instantclient/default.nix:
```diff
@@ -1,71 +1,123 @@
+      "sqlplus" = "0zj5h84ypv4n4678kfix6jih9yakb277l9hc0819iddc0a5slbi5";
+      "odbc"    = "1g1z6pdn76dp440fh49pm8ijfgjazx4cvxdi665fsr62h62xkvch";
+    };
+    "x86_64-darwin" = {
```
added, thanks.

flokli commented on pkgs/development/libraries/oracle-instantclient/default.nix:
```diff
@@ -1,71 +1,123 @@
+
+  # assemble srcs
+  srcs = map (component:
+    (fetcher (srcFilename component arch version "") hashes."${component}" or ""))
```
fixed, thanks.

flokli commented on pkgs/development/libraries/oracle-instantclient/default.nix:
```diff
@@ -1,71 +1,123 @@
-    ln -sfn $out/lib/libclntsh${extLib}.12.1 $out/lib/libclntsh${extLib}
+  postFixup = optionalString stdenv.isDarwin ''
+    for exe in "$out/bin/"* ; do
+      install_name_tool -add_rpath "$lib/lib" "$exe"
```
right :see_no_evil: Thanks!

flokli commented on pkgs/development/libraries/oracle-instantclient/default.nix:
```diff
@@ -1,71 +1,123 @@
+    ".zip";
 
+  # fetcher for the clickthrough artifacts (requiring manual download)
+  fetchClickThrough =  srcFilename: hash: (requireFile {
```
right :see_no_evil: Thanks!

flokli commented:
I adressed all the changes, and could get DBDOracle to build aswell - had to add `${oracle-instantclient.lib}/lib` to the rpath of the resulting `Oracle.bundle` file. The tests now pass there, too.

aanderse reviewed (approved):
Successfully tested connectivity to an `oracle` database using `phpPackages.oci8`. :+1: 

Thanks @flokli!

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/how-do-i-use-cx-oracle-to-connect-to-oracle-db/6944/1
